firedrake/pointwise_operators.py

```diff
@@ -0,0 +1,485 @@
+from abc import ABCMeta, abstractmethod
+from functools import partial
+import types
+from ufl.core.external_operator import ExternalOperator
+from ufl.core.expr import Expr
+from ufl.algorithms.apply_derivatives import VariableRuleset
+from ufl.constantvalue import as_ufl
+from firedrake.function import Function
+from firedrake import utils, functionspaceimpl
+from pyop2.datatypes import ScalarType
+from ufl.log import error
+import sympy as sp
+import numpy as np
+from scipy import optimize
+
+
+class AbstractPointwiseOperator(Function, ExternalOperator, metaclass=ABCMeta):
+    r"""Asbtract base class from wich stem from all the firedrake practical implementation of the
+    ExternalOperator, i.e. all the ExternalOperator subclasses that have mechanisms to be
+    evaluated pointwise and to provide their own derivatives.
+    This class inherits from firedrake.function.Function and ufl.core.external_operator.ExternalOperator
+    Every subclass based on this class must provide the `compute_derivatives` and 'evaluate' methods.
+    """
+
+    def __init__(self, *operands, function_space, derivatives=None, count=None, val=None, name=None, dtype=ScalarType, operator_data=None, extop_id=None):
+        ExternalOperator.__init__(self, *operands, function_space=function_space, derivatives=derivatives, count=count, extop_id=extop_id)
+        aa = function_space
+        ofs = self.original_function_space()
+        if not isinstance(ofs, functionspaceimpl.WithGeometry):
+            fs1 = functionspaceimpl.FunctionSpace(function_space.mesh().topology, ofs.ufl_element())
+            fs = functionspaceimpl.WithGeometry(fs1, function_space.mesh())
+        else:
+            fs = ofs
+
+        Function.__init__(self, fs, val, name, dtype, count=self._count)  # count has been initialised in ExternalOperator.__init__
+
+        self._ufl_function_space = aa
+        self._original_function_space = fs
+        self.operator_data = operator_data
+
+        if extop_id is not None:
+            self.extop_id = extop_id
```

Comment:
What does this do? I am a little scared of the caching implications of operator IDs.

Reply:
It is a deprecated feature that I do not use anymore. I am wondering why I did not remove it, I will do it.